core/src/apps/bitcoin/verification.py

```diff
@@ -0,0 +1,93 @@
+                    raise wire.DataError("Invalid public key hash")
+                return
+            if len(script_pubkey) == 34:  # P2WSH
+                script, self.signatures = read_witness_p2wsh(witness)
+                script_hash = sha256(script).digest()
+                if output_script_native_p2wpkh_or_p2wsh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = read_output_script_multisig(script)
+                return
+        elif witness and witness != b"\x00":
+            if len(script_sig) == 23:  # P2WPKH nested in BIP16 P2SH
+                self.public_keys, self.signatures = read_witness_p2wpkh(witness)
+                pubkey_hash = ecdsa_hash_pubkey(self.public_keys[0], coin)
+                if input_script_p2wpkh_in_p2sh(pubkey_hash) != script_sig:
+                    raise wire.DataError("Invalid public key hash")
+                script_hash = coin.script_hash(script_sig[1:])
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                return
+            if len(script_sig) == 35:  # P2WSH nested in BIP16 P2SH
+                script, self.signatures = read_witness_p2wsh(witness)
+                script_hash = sha256(script).digest()
+                if input_script_p2wsh_in_p2sh(script_hash) != script_sig:
+                    raise wire.DataError("Invalid script hash")
+                script_hash = coin.script_hash(script_sig[1:])
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = read_output_script_multisig(script)
+                return
+        else:
+            if len(script_pubkey) == 25:  # P2PKH
+                self.public_keys, self.signatures = read_input_script_p2pkh(script_sig)
+                pubkey_hash = ecdsa_hash_pubkey(self.public_keys[0], coin)
+                if output_script_p2pkh(pubkey_hash) != script_pubkey:
+                    raise wire.DataError("Invalid public key hash")
+                return
+            if len(script_pubkey) == 23:  # P2SH
+                script, self.signatures = read_input_script_multisig(script_sig)
+                script_hash = coin.script_hash(script)
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = read_output_script_multisig(script)
+                return
+
+        raise wire.DataError("Unsupported signature script")
+
+    def check_hash_type(self, hash_type: int) -> bool:
+        for signature in self.signatures:
+            if signature[1] != hash_type:
+                return False
+        return True
+
+    def verify(self, digest: bytes) -> None:
```

Comment:
This function can be used with any hash type, it is in no way specific to SIGHASH_ALL.

The function does make an assumption that in case of multisig, all signatures are using the same hash type, but it doesn't have to be SIGHASH_ALL.

Reply:
I originally considered checking the assumption that all signatures are using the same hash type in __init__(), but I wanted to make the check optional, because in BIP322 the hash type is actually irrelevant. It's appended to the signature just for historical reasons and the value is not defined. So if different signers happen to append a different hash type to each of their signatures, then it's not a reason to reject the proof.

Reply:
> In that case maybe `hash_type` should be an argument, and the method would (only?) verify matching signatures?

That sounds like a vulnerability waiting to happen.

I think this whole issue is premature. We are not even supporting any other hash types yet, and we are debating support for mixed hash types in multisig. When it comes to that, we can add a `get_hash_types() -> Set[int]` method to the verifier, which would return the set of hash types required. The caller would compute the required ones and then call `verify()`, where the `digest` parameter would be replaced with a dict mapping each hash type to a digest value. An alternative to this solution is to replace the `digest` parameter with a callback `get_digest(hash_type: int) -> bytes`.

Reply:
I'm not interested in going in that direction. This train of thought started with my confusion that the verifier is not tied to a hash_type, despite sort-of needing it to properly verify the signatures.

What I find confusing about this piece of code is that the hash_type validation is separate from signature verification. That seems plainly incorrect.

I am still a little unsure about how hash types work, but currently it seems like this?
- each signature contains a hash_type byte, which is one of SIGHASH_NONE, SIGHASH_SINGLE, SIGHASH_ALL, plus perhaps a flag SIGHASH_ANYONECANSPEND and SIGHASH_FORKID.
- a digest that the signature is signing differs based on the hash_type
- so a validity of the signature does in fact depend on the hash_type; a good signature for a digest might be invalid if the digest is for a mismatched hash_type

Am I missing something?

So. Either we need to take hash_type into account -- and then ensure that the hash-type is propagated with the digest.

Or commit to only supporting SIGHASH_ALL and make that clear from the code, e.g., by rejecting signatures that aren't SIGHASH_ALL.

Or we ignore the existence of hash_types and leave that up to the caller - which *seems* to be what the code is doing now.
But in that case, the code should also make it clear that this is what's happening, e.g., by explicitly throwing away the hash_types.

Reply:
> Am I missing something?

That is correct, plus there will probably be SIGHASH_NOINPUT (BIP 118 draft).

> Or we ignore the existence of hash_types and leave that up to the caller - which seems to be what the code is doing now.

That's my intention to avoid possible compatibility issues in the BIP322 case. As discussed IRL, the solution is to add a code comment explaining this,  rename `check_hash_type()` to `ensure_hash_type()` and as a separate issue to add a txi parameter to `get_hash_type()` for future use.

Reply:
Code comments and renaming of check_hash_type() to ensure_hash_type(): f09d8733c52f0e6105536625e64050153aad402c, 4492c19a42b9b0b691f99ba4cf19e9d527493176,  106a7ab61a88c6159aee155382584ac2ce974783
Add txi parameter: 6431915cce5d57920894a1515cd9b2d3601bfc93